providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/Util.java

```diff
@@ -1,36 +1,123 @@
+     */
+    public static void waitForDesiredState(ManagedChannel ch,
+                                           ConnectivityState desiredState,
+                                           Runnable callback, long timeout,
+                                           TimeUnit unit) throws InterruptedException {
+        waitForDesiredState(ch, desiredState, callback, new CountDownLatch(1), timeout, unit);
+    }
+
+    /**
+     * A recursive helper method to monitor a gRPC channel's state until the desired state is reached or timeout occurs.
+     *
+     * @param ch           the gRPC {@link ManagedChannel} to monitor
+     * @param desiredState the desired {@link ConnectivityState} to wait for
+     * @param callback     a {@link Runnable} to execute when the desired state is reached
+     * @param latch        a {@link CountDownLatch} used for synchronizing the completion of the state change
+     * @param timeout      the maximum time to wait
+     * @param unit         the {@link TimeUnit} for the timeout parameter
+     * @throws InterruptedException if the waiting thread is interrupted
+     * @throws GeneralError         if the deadline is exceeded before reaching the desired state
+     */
+    private static void waitForDesiredState(ManagedChannel ch,
+                                            ConnectivityState desiredState,
+                                            Runnable callback,
+                                            CountDownLatch latch,
+                                            long timeout,
+                                            TimeUnit unit) throws InterruptedException {
+        ch.notifyWhenStateChanged(ch.getState(true), () -> {
+            try {
+                ConnectivityState state = ch.getState(false);
+                log.info("Channel state changed to: {}", state);
+
+                if (state == desiredState) {
+                    callback.run();
+                    latch.countDown();
+                    return;
+                }
+                waitForDesiredState(ch, desiredState, callback, latch, timeout, unit);
+            } catch (Exception e) {
+                log.error("Error during state monitoring", e);
+            }
+        });
+
+        // Await the latch or timeout for the state change
+        if (!latch.await(timeout, unit)) {
+            throw new GeneralError(String.format("Deadline exceeded. Condition did not complete within the %d deadline", timeout));
+        }
+    }
+
+    /**
+     * Monitors the state of a gRPC {@link ManagedChannel} and triggers callbacks for specific state changes.
+     *
+     * @param ch      the gRPC {@link ManagedChannel} to monitor
+     * @param onReady a {@link Runnable} to execute when the channel becomes READY
+     * @param onLost  a {@link Runnable} to execute when the channel enters a TRANSIENT_FAILURE state
+     */
+    public static void monitorChannelState(ManagedChannel ch, Runnable onReady, Runnable onLost) {
```

Comment:
`onLost` can be interpreted as a lot of things, maybe we should rename it to `onConnectionLost` or `onFailure` to signal the connection with the ConnectivityState.